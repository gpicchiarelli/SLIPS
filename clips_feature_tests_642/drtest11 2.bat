(clear)                   ; Optimized Rete Evaluator Issue

(defrule rule-1
   (factoid ?x ?y&:(and ?x ?y)) ; FactPNGetVar3
   =>)

(defrule rule-2
   (factoid ?x ?y)
   (test (and ?x ?y)) ; FactJNGetVar3
   =>)

(defrule rule-3
   (factoid $? ?x ?y&:(and ?x ?y) $?) ; FactPNGetVar1
   =>)

(defrule rule-4
   (factoid $? ?x ?y $?)
   (test (and ?x ?y)) ; FactJNGetVar1
   =>)
(defglobal ?*z* = FALSE)

(defrule rule-5
   (factoid ? ?)
   (test (and ?*z* ?*z*))
   =>)
(assert (factoid FALSE FALSE))
(agenda)
(assert (factoid "FALSE" "FALSE"))
(agenda)
(clear)

(deftemplate factoid
   (slot s1)
   (slot s2))

(defrule rule-1
   (factoid (s1 ?x) (s2 ?y&:(and ?x ?y))) ; FactPNGetVar2
   =>)

(defrule rule-2
   (factoid (s1 ?x) (s2 ?y))
   (test (and ?x ?y)) ; FactJNGetVar2
   =>)
(assert (factoid (s1 FALSE) (s2 FALSE)))
(agenda)
(assert (factoid (s1 "FALSE") (s2 "FALSE")))
(agenda)
(clear)

(defclass OBJOID1
   (is-a USER)
   (slot s1)
   (slot s2))

(defclass OBJOID2
   (is-a USER)
   (multislot ms1))

(defrule rule-1
   (object 
      (is-a OBJOID1)
      (s1 ?x)
      (s2 ?y&:(and ?x ?y))) ; ObjectGetVarPNFunction1
   =>)

(defrule rule-2
   (object 
      (is-a OBJOID1)
      (s1 ?x)
      (s2 ?y))
   (test (and ?x ?y)) ; ObjectGetVarJNFunction1
   =>)

(defrule rule-3
   (object 
      (is-a OBJOID2)
      (ms1 $? ?x ?y&:(and ?x ?y))) ; ObjectGetVarPNFunction2
   =>)
   
(defrule rule-4
   (object 
      (is-a OBJOID2)
      (ms1 $? ?x ?y))
   (test (and ?x ?y)) ; ObjectGetVarJNFunction2
   =>)
(make-instance o1 of OBJOID1 (s1 FALSE) (s2 FALSE))
(make-instance o2 of OBJOID2 (ms1 FALSE FALSE))
(agenda)
(make-instance o3 of OBJOID1 (s1 "FALSE") (s2 "FALSE"))
(make-instance o4 of OBJOID2 (ms1 "FALSE" "FALSE"))
(agenda)
(clear) ; load-facts GC issue

(defglobal MAIN ?*key-id* = 0
                ?*reasmb-id* = 1)

(deffunction key-id ()
   (bind ?*key-id* (+ ?*key-id* 1))
   (bind ?*reasmb-id* 1)
   (return ?*key-id*))
   
(deftemplate key 
   (slot weight (default 1))
   (slot id (default-dynamic (key-id))))
         
(deffunction reasmb-id ()
   (bind ?rv (create$ ?*key-id* 
                      ?*reasmb-id*))
   (bind ?*reasmb-id* (+ ?*reasmb-id* 1))
   (return ?rv))

(deftemplate reasmb
   (multislot id (default-dynamic (reasmb-id))))
(load-facts buglfgc.fct)
(facts)
(clear) ; Calling redefined method body causes crash
(defmethod foo ())
(foo)
(defmethod foo () (bind ?a 1))
(foo)
(clear) ; Composite default-dynamic crash

(defclass FOO
   (is-a USER)
   (slot x (type SYMBOL) 
           (default-dynamic (sym-cat foo bar))))
           
(defclass BAR (is-a USER)
  (slot x (type SYMBOL)))

(defclass WOZ
  (is-a FOO BAR)
  (slot x (type SYMBOL)
          (source composite)))
(make-instance w of WOZ)
(send [w] print)
(clear) ; 6.42 bug
(defglobal ?*foo*  = (create$  0.0)) 

(loop-for-count 1 do
   (bind ?*foo* ?*foo*))
?*foo*
(clear)   
(defglobal ?*x* = (create$ 1 2 3))

(while TRUE do
   (bind ?*x* (rest$ ?*x*))
   (printout t ?*x* crlf)
   (if (= (length$ ?*x*) 0) 
      then (break)))
?*x*
(clear)
(defglobal ?*foo* = (create$ 0.0))

(deffunction test (?foo)
   (bind ?*foo* ?foo)
   (println ?foo)
   (bind ?*foo* (create$ 3.0))
   (println ?foo))
(test ?*foo*)
?*foo*
(clear) ; Modify/duplicate ordered fact
(assert (blah))
(modify 1 (x 3))
(duplicate 1 (y 4))
(clear) ; Missing space in error message

(deftemplate branch
   (slot to))

(deffacts story-flow
   (branch (to obey-the-order Q9)))
   
(defrule error
   =>
   (assert (branch (to obey-the-order Q9))))
(clear) ; Erroneous constraint conflict

(defrule issue
   =>
   (bind ?values "A" "B" "C" "D" "E" "F")
   (bind ?value (nth$ 1 ?values)))
(clear) ; Erroneous constraint conflict

(defrule should-be-undefined-variable-error
   =>
   (println (implode$ ?statement)))
(clear) ; ppdefmodule bug with logical name nil
(ppdefmodule bogus nil)
(clear) ; Crash when generic function call overrides message-handler only functions
(defmethod ppinstance ((?x STRING)))
(defclass FOO (is-a USER))

(defmessage-handler FOO doit()
  (ppinstance))
(make-instance [a] of FOO)
(send [a] doit)
(clear) ; member$ bug

(defrule t1
   (list a ? ? $?nr)
   =>
   (println "t1: " (member$ d ?nr)))
   
(defrule t2
   (list $?nr)
   =>
   (println "t2: " (member$ (create$ b c) ?nr)))
(assert (list a b c d))
(run)
(clear) ; Watch methods
(defmethod myecho ((?x INTEGER)))
(watch methods myecho 1)
(myecho 3)
(unwatch methods myecho 1)
(myecho 3)
(clear) ; Memory leak
(defclass Derived (is-a USER) (pattern-match non-reactive) (slot text))
(make-instance [derived] of Derived)
(defrule test =>)
(make-instance [derived] of Derived)
(clear) ; gensym* bug with make-instance
(setgen 1)
(watch instances)
(defclass POINT (is-a USER) (slot x) (slot y))
(make-instance gen2 of POINT (x 1) (y 2))
(make-instance of POINT)
(make-instance of POINT)
(unwatch instances)
(clear) ; Invalid first character for a variable
(defrule foo (bar ?+) =>)
(defrule foo => (assert (bar ?+)))
(defrule foo (bar $?+) =>)
(defrule foo => (assert (bar $?+)))
(clear)
