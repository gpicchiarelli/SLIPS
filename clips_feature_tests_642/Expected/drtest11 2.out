TRUE
CLIPS> (batch "drtest11.bat")
TRUE
CLIPS> (clear)                   ; Optimized Rete Evaluator Issue
CLIPS> 
(defrule rule-1
   (factoid ?x ?y&:(and ?x ?y)) ; FactPNGetVar3
   =>)
CLIPS> 
(defrule rule-2
   (factoid ?x ?y)
   (test (and ?x ?y)) ; FactJNGetVar3
   =>)
CLIPS> 
(defrule rule-3
   (factoid $? ?x ?y&:(and ?x ?y) $?) ; FactPNGetVar1
   =>)
CLIPS> 
(defrule rule-4
   (factoid $? ?x ?y $?)
   (test (and ?x ?y)) ; FactJNGetVar1
   =>)
CLIPS> (defglobal ?*z* = FALSE)
CLIPS> 
(defrule rule-5
   (factoid ? ?)
   (test (and ?*z* ?*z*))
   =>)
CLIPS> (assert (factoid FALSE FALSE))
<Fact-1>
CLIPS> (agenda)
CLIPS> (assert (factoid "FALSE" "FALSE"))
<Fact-2>
CLIPS> (agenda)
0      rule-1: f-2
0      rule-2: f-2
0      rule-3: f-2
0      rule-4: f-2
For a total of 4 activations.
CLIPS> (clear)
CLIPS> 
(deftemplate factoid
   (slot s1)
   (slot s2))
CLIPS> 
(defrule rule-1
   (factoid (s1 ?x) (s2 ?y&:(and ?x ?y))) ; FactPNGetVar2
   =>)
CLIPS> 
(defrule rule-2
   (factoid (s1 ?x) (s2 ?y))
   (test (and ?x ?y)) ; FactJNGetVar2
   =>)
CLIPS> (assert (factoid (s1 FALSE) (s2 FALSE)))
<Fact-1>
CLIPS> (agenda)
CLIPS> (assert (factoid (s1 "FALSE") (s2 "FALSE")))
<Fact-2>
CLIPS> (agenda)
0      rule-1: f-2
0      rule-2: f-2
For a total of 2 activations.
CLIPS> (clear)
CLIPS> 
(defclass OBJOID1
   (is-a USER)
   (slot s1)
   (slot s2))
CLIPS> 
(defclass OBJOID2
   (is-a USER)
   (multislot ms1))
CLIPS> 
(defrule rule-1
   (object 
      (is-a OBJOID1)
      (s1 ?x)
      (s2 ?y&:(and ?x ?y))) ; ObjectGetVarPNFunction1
   =>)
CLIPS> 
(defrule rule-2
   (object 
      (is-a OBJOID1)
      (s1 ?x)
      (s2 ?y))
   (test (and ?x ?y)) ; ObjectGetVarJNFunction1
   =>)
CLIPS> 
(defrule rule-3
   (object 
      (is-a OBJOID2)
      (ms1 $? ?x ?y&:(and ?x ?y))) ; ObjectGetVarPNFunction2
   =>)
CLIPS>    
(defrule rule-4
   (object 
      (is-a OBJOID2)
      (ms1 $? ?x ?y))
   (test (and ?x ?y)) ; ObjectGetVarJNFunction2
   =>)
CLIPS> (make-instance o1 of OBJOID1 (s1 FALSE) (s2 FALSE))
[o1]
CLIPS> (make-instance o2 of OBJOID2 (ms1 FALSE FALSE))
[o2]
CLIPS> (agenda)
CLIPS> (make-instance o3 of OBJOID1 (s1 "FALSE") (s2 "FALSE"))
[o3]
CLIPS> (make-instance o4 of OBJOID2 (ms1 "FALSE" "FALSE"))
[o4]
CLIPS> (agenda)
0      rule-3: [o4]
0      rule-4: [o4]
0      rule-1: [o3]
0      rule-2: [o3]
For a total of 4 activations.
CLIPS> (clear) ; load-facts GC issue
CLIPS> 
(defglobal MAIN ?*key-id* = 0
                ?*reasmb-id* = 1)
CLIPS> 
(deffunction key-id ()
   (bind ?*key-id* (+ ?*key-id* 1))
   (bind ?*reasmb-id* 1)
   (return ?*key-id*))
CLIPS>    
(deftemplate key 
   (slot weight (default 1))
   (slot id (default-dynamic (key-id))))
CLIPS>          
(deffunction reasmb-id ()
   (bind ?rv (create$ ?*key-id* 
                      ?*reasmb-id*))
   (bind ?*reasmb-id* (+ ?*reasmb-id* 1))
   (return ?rv))
CLIPS> 
(deftemplate reasmb
   (multislot id (default-dynamic (reasmb-id))))
CLIPS> (load-facts buglfgc.fct)
4
CLIPS> (facts)
f-1     (key (weight 1) (id 1))
f-2     (reasmb (id 1 1))
f-3     (reasmb (id 1 2))
f-4     (key (weight 3) (id 2))
For a total of 4 facts.
CLIPS> (clear) ; Calling redefined method body causes crash
CLIPS> (defmethod foo ())
CLIPS> (foo)
FALSE
CLIPS> (defmethod foo () (bind ?a 1))
CLIPS> (foo)
1
CLIPS> (clear) ; Composite default-dynamic crash
CLIPS> 
(defclass FOO
   (is-a USER)
   (slot x (type SYMBOL) 
           (default-dynamic (sym-cat foo bar))))
CLIPS>            
(defclass BAR (is-a USER)
  (slot x (type SYMBOL)))
CLIPS> 
(defclass WOZ
  (is-a FOO BAR)
  (slot x (type SYMBOL)
          (source composite)))
CLIPS> (make-instance w of WOZ)
[w]
CLIPS> (send [w] print)
[w] of WOZ
(x foobar)
CLIPS> (clear) ; 6.42 bug
CLIPS> (defglobal ?*foo*  = (create$  0.0)) 
CLIPS> 
(loop-for-count 1 do
   (bind ?*foo* ?*foo*))
FALSE
CLIPS> ?*foo*
(0.0)
CLIPS> (clear)   
CLIPS> (defglobal ?*x* = (create$ 1 2 3))
CLIPS> 
(while TRUE do
   (bind ?*x* (rest$ ?*x*))
   (printout t ?*x* crlf)
   (if (= (length$ ?*x*) 0) 
      then (break)))
(2 3)
(3)
()
FALSE
CLIPS> ?*x*
()
CLIPS> (clear)
CLIPS> (defglobal ?*foo* = (create$ 0.0))
CLIPS> 
(deffunction test (?foo)
   (bind ?*foo* ?foo)
   (println ?foo)
   (bind ?*foo* (create$ 3.0))
   (println ?foo))
CLIPS> (test ?*foo*)
(0.0)
(0.0)
CLIPS> ?*foo*
(3.0)
CLIPS> (clear) ; Modify/duplicate ordered fact
CLIPS> (assert (blah))
<Fact-1>
CLIPS> (modify 1 (x 3))

[TMPLTFUN2] Ordered facts cannot be used with the 'modify' function.
FALSE
CLIPS> (duplicate 1 (y 4))

[TMPLTFUN2] Ordered facts cannot be used with the 'duplicate' function.
FALSE
CLIPS> (clear) ; Missing space in error message
CLIPS> 
(deftemplate branch
   (slot to))
CLIPS> 
(deffacts story-flow
   (branch (to obey-the-order Q9)))

[TMPLTDEF2] The single field slot 'to' can only contain a single field value.

ERROR:
(deffacts MAIN::story-flow
   (branch (to obey-the-order Q9
CLIPS>    
(defrule error
   =>
   (assert (branch (to obey-the-order Q9))))

[TMPLTDEF2] The single field slot 'to' can only contain a single field value.

ERROR:
(defrule MAIN::error
   =>
   (assert (branch (to obey-the-order Q9
CLIPS> (clear) ; Erroneous constraint conflict
CLIPS> 
(defrule issue
   =>
   (bind ?values "A" "B" "C" "D" "E" "F")
   (bind ?value (nth$ 1 ?values)))
CLIPS> (clear) ; Erroneous constraint conflict
CLIPS> 
(defrule should-be-undefined-variable-error
   =>
   (println (implode$ ?statement)))

[PRCCODE3] Undefined variable ?statement referenced in RHS of defrule.

ERROR:
(defrule MAIN::should-be-undefined-variable-error
   =>
   (println (implode$ ?statement)))
CLIPS> (clear) ; ppdefmodule bug with logical name nil
CLIPS> (ppdefmodule bogus nil)
[PRNTUTIL1] Unable to find defmodule 'bogus'.
""
CLIPS> (clear) ; Crash when generic function call overrides message-handler only functions
CLIPS> (defmethod ppinstance ((?x STRING)))
CLIPS> (defclass FOO (is-a USER))
CLIPS> 
(defmessage-handler FOO doit()
  (ppinstance))
CLIPS> (make-instance [a] of FOO)
[a]
CLIPS> (send [a] doit)
[MSGFUN4] The function 'ppinstance' may only be called from within message-handlers.
[PRCCODE4] Execution halted during the actions of message-handler 'doit' primary in class 'FOO'
FALSE
CLIPS> (clear) ; member$ bug
CLIPS> 
(defrule t1
   (list a ? ? $?nr)
   =>
   (println "t1: " (member$ d ?nr)))
CLIPS>    
(defrule t2
   (list $?nr)
   =>
   (println "t2: " (member$ (create$ b c) ?nr)))
CLIPS> (assert (list a b c d))
<Fact-1>
CLIPS> (run)
t1: 1
t2: (2 3)
CLIPS> (clear) ; Watch methods
CLIPS> (defmethod myecho ((?x INTEGER)))
CLIPS> (watch methods myecho 1)
CLIPS> (myecho 3)
MTH >> myecho:#1  ED:1 (3)
MTH << myecho:#1  ED:1 (3)
FALSE
CLIPS> (unwatch methods myecho 1)
CLIPS> (myecho 3)
FALSE
CLIPS> (clear) ; Memory leak
CLIPS> (defclass Derived (is-a USER) (pattern-match non-reactive) (slot text))
CLIPS> (make-instance [derived] of Derived)
[derived]
CLIPS> (defrule test =>)
CLIPS> (make-instance [derived] of Derived)
[derived]
CLIPS> (clear) ; gensym* bug with make-instance
CLIPS> (setgen 1)
1
CLIPS> (watch instances)
CLIPS> (defclass POINT (is-a USER) (slot x) (slot y))
CLIPS> (make-instance gen2 of POINT (x 1) (y 2))
==> instance [gen2] of POINT
[gen2]
CLIPS> (make-instance of POINT)
==> instance [gen1] of POINT
[gen1]
CLIPS> (make-instance of POINT)
==> instance [gen3] of POINT
[gen3]
CLIPS> (unwatch instances)
CLIPS> (clear) ; Invalid first character for a variable
CLIPS> (defrule foo (bar ?+) =>)

[SCANNER2] The character '+' is not valid as the first character of a variable name.

[PRNTUTIL2] Syntax Error:  Check appropriate syntax for defrule.

ERROR:
(defrule MAIN::foo
   (bar ?+
CLIPS> (defrule foo => (assert (bar ?+)))

[SCANNER2] The character '+' is not valid as the first character of a variable name.

[PRNTUTIL2] Syntax Error:  Check appropriate syntax for RHS patterns.

ERROR:
(defrule MAIN::foo
   =>
   (assert (bar ?+
CLIPS> (defrule foo (bar $?+) =>)

[SCANNER2] The character '+' is not valid as the first character of a variable name.

[PRNTUTIL2] Syntax Error:  Check appropriate syntax for defrule.

ERROR:
(defrule MAIN::foo
   (bar $?+
CLIPS> (defrule foo => (assert (bar $?+)))

[SCANNER2] The character '+' is not valid as the first character of a variable name.

[PRNTUTIL2] Syntax Error:  Check appropriate syntax for RHS patterns.

ERROR:
(defrule MAIN::foo
   =>
   (assert (bar $?+
CLIPS> (clear)
CLIPS> (dribble-off)
